### ISSUE TITLE
Rooms need to be in the Conference object

### ISSUE THREAD

**Issue by BenignCremator** [bethlakshmi/GBE2#584]
The room reports need to be able to select conference to show the correct room assignments.  However, we may or may not have the same rooms from year to year, so the rooms should be in the Conference, so bookings are by Conference (or day of a Conference) instead of being a disassociated date that we then need to winnow by Conference date.  GBE-569 is blocked on this.

**Comment by jonkiparsky** [comment]
Simple enough change to make.
1) Add foreign key(Conference) to Room object
2) create schemamigration, default to GBE9
3) create datamigration to (a) duplicate existing rooms with GBE10 and (b) associate GBE10 events with GBE10 rooms

**Comment by jonkiparsky** [comment]
I can take this item after current work on #488

**Comment by bethlakshmi** [comment]
I don't agree.  Maybe we can discuss on Thursday?

A room exists whether or not there is a conference.  I don't see why we need to do the extra work to make different rooms for each year of the conference.  In fact, there may be something valuable to knowing the given room's history across the conferences.

The room report *does* have to be filtered by conference, but I don't see why we can't
- choose from the list of conferences in the system
- select events relevant to a chosen conference for each room.

There's more to the change than what Jon writes, we also have to go into the event creation/scheduling logic and update it so that the user is associating the room for this conference with this event. 

I'm just not sure that I understand why the added complexity is valuable here.

**Comment by jonkiparsky** [comment]
Doing things wrong to save time has fucked this month for me. Not continuing in this vein if we can help it.

**Comment by BenignCremator** [comment]
On 11/30/2015 09:19 AM, bethlakshmi wrote:

> I don't agree. Maybe we can discuss on Thursday?

If we wait until Thursday, Jon will not have time to make the changes to
conferences, and even if he did, I would not have time to make the
changes to all of the reports by the Dec 4th deadline.  And there will
be no time to test the code before deploy.  Right now, not being able to
retrieve data by conferences is breaking critical functionality on the 
website that is needed in Dec.

> A room exists whether or not there is a conference. I don't see why
> we need to do the extra work to make different rooms for each year
> of the conference.

Because it will simplify the views code immeasurably.  Once things that 
get scheduled in a conference are part of a conference, I can get them 
out of the database with a single call.  However, if we do not make 
rooms part of conferences, we then have to do completely independent 
database calls and iterate over the results to winnow out the ones that 
are not part of this conference.  So that is two or three database calls
compared to one, plus winnowing code, because we have no way to perform
joins on the database.  That is the real crux of the problem.  We talk
about putting rooms into the conferences, because that is how we
structure the code, however, Python "compiles" that down to separate
tables that have indexes for rooms within the conference tables, which
makes searches quick and easy.  Versus having separate tables without
indexes, which is what having rooms not part of conferences compiles
into.  Worse, the things that we would match on would be generated 
separately, so the matching code has to work around the change in format 
that can create.  This is more or less what made bookings such a mess of 
code last year that had me pulling my hair out and not wanting to commit 
to coding so much.  And we would have to do that everywhere we want to 
match sets of data together, which is nearly everywhere.

> In fact, there may be something valuable to knowing the given room's
> history across the conferences.

This can be done by not filtering by the conference slug when we call 
the conference objects from the database, and iterating over that for 
matches.  And yes, that is somewhat messy and requires hand tweaked 
code, but it is better to do that for something one might use 
occasionally then to do it for everything else that you use all of the time.

> The room report /does/ have to be filtered by conference, but I
> don't see why we can't

> * choose from the list of conferences in the system
 > * select events relevant to a chosen conference for each room.

...  That is how the UI will work, either way.  What we are talking
about here is how the underlying infrastructure will work once we make
the selection.  Completely different parts of the site.

Hunter

**Comment by bethlakshmi** [comment]
First off, when I wrote #569  - it specifically said "before the conference".  The room reports are not used until the week before the conference, and I wrote very specifically in a comment that this was "medium" until the month before the conference.

Please let's not try to bang this out quickly before Dec. 5 if there is other work that is actually tagged for the Dec. 5 milestone!  For reference, I actually set up that milestone weeks ago:

https://github.com/bethlakshmi/GBE2/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Dec+5th+2015%22

So... I don't buy that speed is a good reason not to talk this through.

With that said - if it's simplifying the DB access tremendously - I'm tentatively OK with it... but I will go on record with disliking the idea that we're creating a fake context because it seems to be the easy way.  To me, the room report *is* an edge case - it's a piece of code that tries to link a bunch of things together for a specific use for 1 specific person - to the best of my knowledge, the only person who uses that report is me, I print out the room labels 1-2 times (usually something gets rebooked during the conference so I may print more than 1 copy).  Scandal uses the room setup report once and that's it.  

Trying to think about what a room actually is - it's a structure in the physical world that is independant of the conference.  If we had two conferences running at the same time - we would still only have 1 room.  Regardless of whether we'd booked the room in two different conferences, we'd want to know that we booked the room for two events at the same time.

When @BenignCremator says "This can be done by not filtering by the conference slug when we call the conference objects from the database, and iterating over that for matches."

So... that means string matches.  And we recreate the room every year, right?  And so if the human recreates the room with a spelling mistake in the name, the code breaks?  This seems fragile. 

Up sides to room per conference:
- it simplifies code in the room reports, which are rarely run
- it *may* simplify the calendar code (can we check on that?) which is run a LOT

Down sides to room per conference:
- we have to refactor the event scheduling logic 
- we have to make sure that rooms get re-created for each conference (either a script or more manual work) - if rooms ever get more data attributes (right now they are simple, but if we have other descriptors like "available layout", or "mirrors y/n" or floor surface or whatever), then those have to be recopied every event.
- we have to make sure that when we compare room bookings for conflicts or to check a room's history of bookings, we collect and compare all rooms with the same name.  I could actually see this coming up a lot more than in a single edge case.

My biggest concern is that it breaks the simulation of the real world that I thought we were shooting for with scheduling - when we crafted resource allocations of workers, rooms, and stuff it was my take that this was stuff that was persistent in the real world and available for use in a given event.  Conferences are collections that represent sets of events and bids to do things for the conference, but they are largely centered around something that happens for finite place and time.

**Comment by bethlakshmi** [comment]
Biggest thought from that wall of text... 

what is the right answer w/respect to the calendars?  Right now the calendars only show 2015 events to me.  I'm betting that Friday/Saturday/Sunday are hardcoded in there somewhere.

but if it's a mess to try to pull rooms for Calendars w/out having a room per conference approach - then I'm on board with making rooms unique to conferences.  Then 2 different parts of the code need the same thing for 2 different reasons and I can get over the fact that this isn't a good simulation of the real world.

**Comment by jonkiparsky** [comment]
> we have to refactor the event scheduling logic
Trivially. 

> we have to make sure that rooms get re-created for each conference (either a script or more manual work) - if rooms ever get more data attributes (right now they are simple, but if we have other descriptors like "available layout", or "mirrors y/n" or floor surface or whatever), then those have to be recopied every event.

We've managed to clone biddables without much trouble. This is also trivial


> we have to make sure that when we compare room bookings for conflicts or to check a room's history of bookings, we collect and compare all rooms with the same name. I could actually see this coming up a lot more than in a single edge case.

We have no provisions for this in the current system, and there's been no suggestion that this is something anyone has ever wanted to do. Is this something that someone wants, and just hasn't mentioned?

**Comment by jonkiparsky** [comment]
> this isn't a good simulation of the real world.

Okay, so the simple fact is that in order for this to work as intended, we have to be able to represent the current set of rooms that are available to the conference. Conferences are not always held in the same place year on year, and when they are they do not always have the same rooms year on year. This is the reality we're modeling - not the special case that happens to be true last year and this year.

Given that, I see two ways of doing this. One way is to create a set of rooms for each iteration of the conference. The other way would be to create an M2M between Room and Conference. The former is preferable, because it avoids aliasing issues, meaning that it's simple to change details of a room (booking size, name, etc) from one year to the next without altering previous conference state. 

The issue of identity of room across time is an interesting question. If we need to represent that for some reason, there are a few viable solutions, but none of them are both simple and good. Probably the simplest would be to maintain a "cloned_from" field on a room, and just walk back through time. A more robust and complicated solution would involve creating a Platonic "Crispus Attucks", and locating actual events in shadows of that ideal Room. This seems like overkill to me, though.

**Comment by bethlakshmi** [comment]
For the record- Jon - you and I mentioned checking resource conflicts w/in
the conference when we whiteboarded the scheduling design this time last
year.  But we never wrote a ticket.

I'm willing to agree that cross conference checking is a long shot.  I'm
still interested in what the best approach for calendars is.

**Comment by bethlakshmi** [comment]
Done.  Did it last year.
